src/ir/subtypes.h

```diff
@@ -62,6 +66,45 @@ struct SubTypes {
     return ret;
   }
 
+  // Efficiently traverse subtypes of a type, up to a particular depth (depth =
+  // 0 means not to traverse subtypes, etc.). The callback function receives
+  // (type, depth).
+  template<typename F>
+  void traverseSubTypes(HeapType type, Index depth, F func) {
+    // Start by traversing the type itself.
+    func(type, 0);
+
+    // getStrictSubTypes() returns vectors of subtypes, so for efficiency store
+    // pointers to those in our work queue to avoid allocations. See the note
+    // below on typeSubTypes for why this is safe.
+    struct Item {
+      const std::vector<HeapType>* vec;
+      Index depth;
+    };
+
+    // Real-world type hierarchies tend to have a limited depth, so try to avoid
+    // allocations in our work queue with a SmallVector.
+    SmallVector<Item, 10> work;
+
+    // Start with the subtypes of the base type. Those have depth 1.
+    work.push_back({&getStrictSubTypes(type), 1});
+
+    while (!work.empty()) {
+      auto& item = work.back();
+      work.pop_back();
+      auto currDepth = item.depth;
+      auto& vec = *item.vec;
+      if (currDepth > depth || vec.empty()) {
+        // Nothing we need to traverse here.
+        continue;
+      }
```

Comment:
Slightly, perhaps, but the code would be more verbose as we'd need to add a check before the loop (for the initial work). The current code unifies into a single check location. But, maybe it's worth the extra code, since it's faster... I'll refactor.

Reply:
Yeah, that's probably better, thanks. Done.